### PR TITLE
Stitch1D: respect user EndOverlap value

### LINKED ISSUE #22921: Refine Stitch1D for point data
Sometimes, point data which do not overlap shall be stitched in order to simplify the reduction output, i.e. one workspace. In particular, Stitch1D throws in that case a RuntimeError
```
Stitch1D-[Error] Error in execution of algorithm Stitch1D:
Stitch1D-[Error] Stitch1D cannot have a StartOverlap > EndOverlap. StartOverlap: 0.052642825, EndOverlap: 0.044180824
Stitch1DMany-[Error] Error in execution of algorithm Stitch1DMany:
Stitch1DMany-[Error] Stitch1D cannot have a StartOverlap > EndOverlap. StartOverlap: 0.052642825, EndOverlap: 0.044180824
```

Maybe there is a way to only report the error once and not twice.

The corresponding image is

comp

The actual work around is (simplified)

```
try:
    reduced = mtd[redSet[0]]
except KeyError:
    try:
        Scale(InputWorkspace=refl2, OutputWorkspace=refl2, Factor=factor)
        reduced = ConjoinXRuns(InputWorkspaces=[refl1, refl2])
    except ValueError:
        reduced = mtd[refl1]
```

### PR DESCRIPTION
**Pull request by vreimund**
**Description of work.**

- It seemed to exist an implementation error (use of binIndexOf) to use auto- generated `EndOverlap` values for binned data. The binIndexOf was critical, tolerance and index management was responsible.

- The scaling of the wrong workspace seemed to happen.

- Stitch1DMany has only one set of rebin params and not defining the overlaps would not produce expected results (no documentation): now, the rebin step is used to stitch workspaces pairwise and a final rebinning is performed using `Params`. Otherwise, the rebinning of the first workspaces define already not sensible overlap values for the next workspaces to stitch.

The above mentioned errors seem to be present since some time - tests may break.

Some code changes should now simplify the whole execution and make it easier to maintain and read, hopefully.
- A new template will be able to deliver the start and end overlap values.
- Some max_element, min_element was removed since binned data are sorted and not sorted point data are not supported.
- Repeated `Stitch1D` mentioning in log messages was deleted.
- More input validation to prevent unexpected results (Stitch1DMany using reasonable start and end overlaps controlled by Stitch1D)

It would be nice to add some units to the logging of overlap values since they are x values with units.
And it would be easy to not only recognize special values but also masked bins.

Enables stitching of non-overlapping point data.

Please carefully check the release notes (framework and/or (?) reflectometry), wrongly applied scaling, automatic `EndOverlap` value.

**To test:**

**Attention**: unit tests are modified! Please make sure that EndOverlap was wrongly applied. Should be finally tested with ISIS reflectometry data.

Fixes #22921 and #22851.

Release notes. 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by gemmaguest** [comment]
I'm confused by the new results for the histogram example in the linked issue. For the `ws_` output (which I've renamed `overlap_4_to_5` in the script below) it has an overlap region of `x=4` to `x=5`. The value at `x=4.5` was `7.5` in the old code but in the new code it is `5`. Since this is inside the overlap region, shouldn't it still be `7.5`?

Before:
stitch_result_before

After:
stitch_result_after

```
CreateWorkspace(OutputWorkspace='ws1', DataX='0,1,2,3,4,5,6,7', DataY='5,5,5,5,5,5,5')
CreateWorkspace(OutputWorkspace='ws2', DataX='2,3,4,5,6,7,8,9', DataY='5,5,5,5,5,5,5')
Scale(InputWorkspace='ws1',Factor=2,OutputWorkspace='ws1_scaled')
#Stitch1D(LHSWorkspace='ws1', RHSWorkspace='ws2', OutputWorkspace='overlap_2_to_max', StartOverlap=2, ScaleRHSWorkspace=False, UseManualScaleFactor=True, ManualScaleFactor=2)
Stitch1D(LHSWorkspace='ws1', RHSWorkspace='ws2', OutputWorkspace='overlap_4_to_5', StartOverlap=4, EndOverlap=5, ScaleRHSWorkspace=False, UseManualScaleFactor=True, ManualScaleFactor=2)
#Stitch1D(LHSWorkspace='ws1', RHSWorkspace='ws2', OutputWorkspace='overlap_4_to_6', StartOverlap=4, EndOverlap=6, ScaleRHSWorkspace=False, UseManualScaleFactor=True, ManualScaleFactor=2)
#Stitch1D(LHSWorkspace='ws1', RHSWorkspace='ws2', OutputWorkspace='overlap_4_to_max', StartOverlap=4, ScaleRHSWorkspace=False, UseManualScaleFactor=True, ManualScaleFactor=2)
```

**Comment by reimundILL** [comment]
While I do agree,:+1:, that the fix is not valid, I still think there is a bug.
Here my thinking about expected behavior and master behavior:

```
CreateWorkspace(OutputWorkspace='ws1', DataX='0,1,2,3,4,5,6,7', DataY='5,5,5,5,5,5,5')
CreateWorkspace(OutputWorkspace='ws2', DataX='2,3,4,5,6,7,8,9', DataY='5,5,5,5,5,5,5')

Minimum overlap is 2
Maximum overlap is 7

--
StartOverlap = 2
EndOverlap = 7

Bin edges: 0    1    2     3     4     5     6     7    8   9
Y values:    10   10   7.5   7.5   7.5   7.5   7.5   5    5     MY EXPECTATION
MASTER:      10   10   7.5   7.5   7.5   7.5    5    5    5     WRONG

--

StartOverlap = 4
EndOverlap = 5

Bin edges: 0    1    2    3    4     5    6   7   8   9
Y values:    10   10   10   10   7.5    5   5   5   5
MASTER OK
--

StartOverlap = 4
EndOverlap = 6

Bin edges: 0    1    2    3    4     5     6   7   8   9
Y values:    10   10   10   10   7.5   7.5   5   5   5
MASTER OK
--

StartOverlap = 4
EndOverlap = 7

Bin edges: 0    1    2    3    4     5     6     7   8   9
Y values:    10   10   10   10   7.5   7.5   7.5   5   5        MY EXPECTATION
MASTER       10   10   10   10   7.5   7.5    5    5   5        WRONG
```

I will move on fixing this ... hope you agree on the expected bahavior.

**Comment by reimundILL** [comment]
One more thing to think about: the rhs should be scaled by default. It looks here that the lhs gets scaled:
Switch of `10`s and `5`s in all examples.

I found a solution which shows this and push after some more tests.

I implemented a template in the header to return overlap limits and moved much overlap limit testing to the input validation. Code becomes simpler and easier to understand and maintain. This is my hope.

**Comment by reimundILL** [comment]
Unit tests break now due to scaling which ws (left or right) and new end bin (use of `binIndexOf` -> increase index of endOverlap):

[Stitch.ipynb.zip](https://github.com/mantidproject/mantid/files/2588995/Stitch.ipynb.zip)

Values of `StartOverlap` and `EndOverlap` which are reported are the same for this change and on master.

I plan to add the given tests as additional unit tests and fix the existing tests, if this implementation is ok from your side.

**Comment by reimundILL** [comment]
Hi Gemma, this branch is ready now for detailed testing. All my tests seem to be successful, I will add them as new unit tests and fix the wrong unit tests. Then only new system tests are needed which I cannot generate.

**Comment by reimundILL** [comment]
I needed now apply a global fix for the indices and some more constness. The existing tests should not fail.

**Comment by reimundILL** [comment]
One test (test_stitches_three_no_overlaps_specified_should_still_work) fails in `Stitch1DMany` which passed before, here the explanation:
- Params gives the rebin parameters for three to stitch workspaces this causes accuracy problems in the range. (last bin boundary out of range by 1.e-6). I was not able to include this tolerance since other tests fail then.
- A very detailed log message will now point out what do do: it may make more sense to give only a rebin step (which will be in the test now). Alternatively, `EndOverlaps` can be defined to overcome this tolerance problem.
This is not too much to ask from the user. I think the user would expect that the two first workspaces determine the first set of overlaps and the second and third workspaces determine the second set of overlap values automatically. BUT when the rebin range is the final expected range, the lhs of the second stitch session has complete range while the rhs is part of this range. Overlap last value will be 2.6 (**+1.e-6**) instead of something expected like 1.8! I observed that by using `getBinIndexOf` this fails due to bin boundary out of workspace range.
In conclusion: everything depends on bin boundary tolerances (1.e-9) and something like rebin may mess this up.

**Comment by gemmaguest** [comment]
I'm getting a crash with a very small overlap region:
```
CreateWorkspace(OutputWorkspace='ws1', DataX='0,1,2,3,4,5,6,7', DataY='5,5,5,5,5,5,5')
CreateWorkspace(OutputWorkspace='ws2', DataX='2,3,4,5,6,7,8,9', DataY='5,5,5,5,5,5,5')
Stitch1D(LHSWorkspace='ws1', RHSWorkspace='ws2', OutputWorkspace='overlap_small', StartOverlap=4.1, EndOverlap=4.2, ScaleRHSWorkspace=False, UseManualScaleFactor=True, ManualScaleFactor=2)
```

**Comment by reimundILL** [comment]
I will add a test for this case in `Stitch1D`.

**Comment by reimundILL** [comment]
For `CreateTransmissionWorkspace2Test`:
The rebinning params are {1.7924071518777369, 0.89620357593886846, 14.339257215021895}.
The start and end overlaps are not given -> will be determined.
The determined end overlap was 14.339257217021895 and thus out of range.

While the start overlap is never smaller than the smallest bin bounday (since + params[1]), the end overlap could exceed the last bin boundary. I now take the last bin index for those cases. The failing tests pass for me now.

**Comment by reimundILL** [comment]
Rebin parameters of the failing test on master: {0.011088383560452199, -0.040000000000000001, 0.83179794140121199} look insane. There is a need of rebin parameter checking.

**Comment by gemmaguest** [comment]
With a small overlap region for point data I get a divide-by-zero:

```
CreateWorkspace(OutputWorkspace='ws1', DataX='0,1,2,3,4,5,6,7', DataY='5,5,5,5,5,5,5,5')
CreateWorkspace(OutputWorkspace='ws2', DataX='2,3,4,5,6,7,8,9', DataY='5,5,5,5,5,5,5,5')
Scale(InputWorkspace='ws1',Factor=2,OutputWorkspace='ws1_scaled')
Stitch1D(LHSWorkspace='ws1_scaled', RHSWorkspace='ws2', OutputWorkspace='overlap_small', StartOverlap=4, EndOverlap=4.1, ScaleRHSWorkspace=False, UseManualScaleFactor=False)
```
Output:
```
Division by zero: the RHS is a single-valued vector with value zero.
Calculated scale factor is: -nan. Check that in both input workspaces the integrated overlap region is non-zero.
```

**Comment by reimundILL** [comment]
Maybe the last commit is not valid? If you do not have an overlap with point data, you cannot stitch, or? The algorithm may want to throw. As well as for binned data.

On master, you get the mentioned behavior of zero scale factors. For both, point and binned data.

I plan to revert the last change that I just did, or?

**Comment by reimundILL** [comment]
There are two cases which differ for automatic scaling:
- the workspaces have an overlap region or not
- the workspaces have an overlap region but the user defines a too small overlap (within one bin, between two points)

For manual scaling this may not be critical:
- for non overlapping point data
- for binned data (here, we rebin to a Stitch1D full range of lhs and rhs and do scaling afterwards)

**Comment by reimundILL** [comment]
It can be solved by input validation but is a bit tricky:

User gives StartOverlap and EndOverlap and wants autmomatic scaling, then we do:

point data: check for all points, that start and end overlap covers minimum 2 points
binned data: check that end - start overlap is larger than bin width or automatic determined bin width
problem binned data: Params has more than three values: do it similar to point data

This is more complicated but trains the user.

**Comment by reimundILL** [comment]
A divide-by-zero (possible on master build) will not be possible anymore.

**Comment by gemmaguest** [comment]
@reimundILL I'm looking into the system test failure for `INTERReductionTest`. I think the differences in the result of the Fit are because of differences in the stitched workspace `44984_85`. Here's a cut-down version of the script that creates this stitched workspace:

```
Load(Filename='INTER00044986.raw', OutputWorkspace='44986', LoaderName='LoadRaw', LoaderVersion='3')
Load(Filename='INTER00044987.raw', OutputWorkspace='44987', LoaderName='LoadRaw', LoaderVersion='3')
Load(Filename='INTER00044988.raw', OutputWorkspace='44988', LoaderName='LoadRaw', LoaderVersion='3')
Load(Filename='INTER00044989.raw', OutputWorkspace='44989', LoaderName='LoadRaw', LoaderVersion='3')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44988', SecondTransmissionRun='44989', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_SM')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44986', SecondTransmissionRun='44987', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_NoSM')
Load(Filename='INTER00044984.raw', OutputWorkspace='44984.raw', LoaderName='LoadRaw', LoaderVersion='3')
Load(Filename='INTER00044985.raw', OutputWorkspace='44985.raw', LoaderName='LoadRaw', LoaderVersion='3')
ReflectometryReductionOneAuto(InputWorkspace='44984.raw', ThetaIn='0.80000000000000004', WavelengthMin='2.6000000000000001', FirstTransmissionRun='TRANS_SM', MomentumTransferMin='0.010327523398054171', MomentumTransferStep='0.055042629842549297', MomentumTransferMax='0.067186475747202865', ScaleFactor='1', OutputWorkspaceBinned='44984_IvsQ_binned', OutputWorkspace='44984_IvsQ', OutputWorkspaceWavelength='44984_IvsLam')
ReflectometryReductionOneAuto(InputWorkspace='44985.raw', ThetaIn='2.2999999999999998', FirstTransmissionRun='TRANS_noSM', MomentumTransferMin='0.029696022073279912', MomentumTransferStep='0.055038929363870739', MomentumTransferMax='0.33476774783608887', ScaleFactor='1', OutputWorkspaceBinned='44985_IvsQ_binned', OutputWorkspace='44985_IvsQ', OutputWorkspaceWavelength='44985_IvsLam')
Stitch1DMany(InputWorkspaces='44984_IvsQ_binned,44985_IvsQ_binned', OutputWorkspace='44984_85', Params='-0.0550419', OutScaleFactors='1.20902')
```

On the branch for this PR, I get a spike in the stitched output just before the start of the overlap range:
stitch_44984_85_new

Whereas in the nightly build or on master it is smooth:
stitch_44984_85_old

Do you know what is causing the difference?

**Comment by reimundILL** [comment]
Thank you, I like to run this example to investigate but have troubles finding the data. SystemTestData I guess. There is a folder INTER containing `INTERReductionTestRuns.nxs`... Can you point me to the data?

**Comment by gemmaguest** [comment]
Sorry, I was loading them from the archive, but they are all in the file you mention. This should work if `<build_directory>/ExternalData/Testing/Data/SystemTest/INTER` is in your user directories:
```
Load(Filename='INTERReductionTestRuns', OutputWorkspace='runs')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44988', SecondTransmissionRun='44989', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_SM')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44986', SecondTransmissionRun='44987', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_NoSM')
ReflectometryReductionOneAuto(InputWorkspace='44984.raw', ThetaIn='0.80000000000000004', WavelengthMin='2.6000000000000001', FirstTransmissionRun='TRANS_SM', MomentumTransferMin='0.010327523398054171', MomentumTransferStep='0.055042629842549297', MomentumTransferMax='0.067186475747202865', ScaleFactor='1', OutputWorkspaceBinned='44984_IvsQ_binned', OutputWorkspace='44984_IvsQ', OutputWorkspaceWavelength='44984_IvsLam')
ReflectometryReductionOneAuto(InputWorkspace='44985.raw', ThetaIn='2.2999999999999998', FirstTransmissionRun='TRANS_noSM', MomentumTransferMin='0.029696022073279912', MomentumTransferStep='0.055038929363870739', MomentumTransferMax='0.33476774783608887', ScaleFactor='1', OutputWorkspaceBinned='44985_IvsQ_binned', OutputWorkspace='44985_IvsQ', OutputWorkspaceWavelength='44985_IvsLam')
Stitch1DMany(InputWorkspaces='44984_IvsQ_binned,44985_IvsQ_binned', OutputWorkspace='44984_85', Params='-0.0550419', OutScaleFactors='1.20902')
```

**Comment by reimundILL** [comment]
My intermediate findings comparing master/branch:

TRANS_SM : Log mismatch (I investigated sample logs, did not yet see the difference) and no data mismatch

TRANS_NoSM: data mismatch due to computational accuracy:
(bin 0) Difference (X,Y,E) = (0,3.66374e-15,2.90566e-17)
(bin 1) Difference (X,Y,E) = (4.44089e-16,3.80251e-15,2.9924e-17)
(bin 2) Difference (X,Y,E) = (8.88178e-16,3.63598e-15,2.81893e-17)
(bin 3) Difference (X,Y,E) = (1.33227e-15,3.94129e-15,3.1225e-17)
(bin 4) Difference (X,Y,E) = (1.77636e-15,4.32987e-15,3.42608e-17)
(bin 5) Difference (X,Y,E) = (2.22045e-15,4.13558e-15,3.20924e-17)
(bin 6) Difference (X,Y,E) = (2.66454e-15,4.2466e-15,3.29597e-17)
(bin 7) Difference (X,Y,E) = (3.10862e-15,5.71765e-15,4.42354e-17)
(bin 8) Difference (X,Y,E) = (3.55271e-15,4.35763e-15,3.25261e-17)
(bin 9) Difference (X,Y,E) = (3.9968e-15,4.69069e-15,3.55618e-17)
(bin 10) Difference (X,Y,E) = (4.44089e-15,3.19189e-15,2.29851e-17)
no outlier observed

44984_85: data mismatch with one bin of significant difference:
(bin 10) Difference (X,Y,E) = (0,5.55112e-17,4.33681e-19)
(bin 11) Difference (X,Y,E) = (0,2.77556e-17,0)
(bin 13) Difference (X,Y,E) = (0,6.93889e-18,0)
(bin 18) Difference (X,Y,E) = (0,0,1.35525e-20)
**(bin 19) Difference (X,Y,E) = (0,0.00374557,3.52225e-05)**
(bin 20) Difference (X,Y,E) = (0,1.04083e-17,2.57498e-19)
(bin 21) Difference (X,Y,E) = (0,4.33681e-18,1.49078e-19)
(bin 22) Difference (X,Y,E) = (0,6.07153e-18,1.35525e-20)
(bin 23) Difference (X,Y,E) = (0,1.30104e-18,6.09864e-20)
(bin 24) Difference (X,Y,E) = (0,4.33681e-19,2.71051e-20)

**Comment by reimundILL** [comment]
Would help here:
https://github.com/mantidproject/mantid/pull/24553

**Comment by NickDraper** [comment]
Conflicts the limerick
```
There once was a branch with conflicts.
Delays to the PR this predicts.
His colleague urged
with master he merged,
and then those conflicts he could fix.
```

**Comment by martyngigg** [comment]
This is currently marked for release 4.0. Should it still be there?

**Comment by reimundILL** [comment]
I will have a look at it.

**Comment by gemmaguest** [comment]
I've had another look at the spike in the stitched curve for `44984_85`. I wonder if it's because the rebin uses the LHS min and the RHS max to rebin both inputs. This means that in the rebinned RHS workspace (`44985_IvsQ_rebin_for_stitch`), the first bin containing a non-zero value has been averaged with the adjacent bin containing a zero value.

It looks like this could cause the downward spike in the stitched result? I think this would make sense with the changes, because they are now inclusive of the point at the edge of the overlap region, whereas before this was excluded?

stitch1d_rebin_inputs_logarithmic_binning

Here's the script I'm using to create these:
```
Load(Filename='INTERReductionTestRuns', OutputWorkspace='runs')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44988', SecondTransmissionRun='44989', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_SM')
CreateTransmissionWorkspaceAuto(FirstTransmissionRun='44986', SecondTransmissionRun='44987', StartOverlap='10', EndOverlap='12', OutputWorkspace='TRANS_NoSM')
ReflectometryReductionOneAuto(InputWorkspace='44984.raw', ThetaIn='0.80000000000000004', WavelengthMin='2.6000000000000001', FirstTransmissionRun='TRANS_SM', MomentumTransferMin='0.010327523398054171', MomentumTransferStep='0.055042629842549297', MomentumTransferMax='0.067186475747202865', ScaleFactor='1', OutputWorkspaceBinned='44984_IvsQ_binned', OutputWorkspace='44984_IvsQ', OutputWorkspaceWavelength='44984_IvsLam')
ReflectometryReductionOneAuto(InputWorkspace='44985.raw', ThetaIn='2.2999999999999998', FirstTransmissionRun='TRANS_noSM', MomentumTransferMin='0.029696022073279912', MomentumTransferStep='0.055038929363870739', MomentumTransferMax='0.33476774783608887', ScaleFactor='1', OutputWorkspaceBinned='44985_IvsQ_binned', OutputWorkspace='44985_IvsQ', OutputWorkspaceWavelength='44985_IvsLam')
Stitch1DMany(InputWorkspaces='44984_IvsQ,44985_IvsQ', OutputWorkspace='44984_85', Params='-0.0550419', OutScaleFactors='1.20902')

# Manually perform a rebin of the inputs using LHS min X and RHS max X and a logarithmic binning step
lhs=mtd['44984_IvsQ']
rhs=mtd['44985_IvsQ']
min=lhs.readX(0)[0]
max=rhs.readX(0)[-1]
step=-0.055
params = str(min) + ',' + str(step) + ','+str(max)
print params

Rebin(InputWorkspace='44984_IvsQ', OutputWorkspace='44984_IvsQ_rebin_for_stitch', Params=params, PreserveEvents=False, FullBinsOnly=True)
Rebin(InputWorkspace='44985_IvsQ', OutputWorkspace='44985_IvsQ_rebin_for_stitch', Params=params, PreserveEvents=False, FullBinsOnly=True)
```

**Comment by NickDraper** [comment]
@gvardany  
Verena has left now.  @gemmaguest is this something the @alicerussell1 could pick up as the developer?

**Comment by gemmaguest** [comment]
@NickDraper @gvardany This was driven by requests from scientists at ILL and these aren't problems we've noticed at ISIS (we're actually quite concerned about making changes to stitching), so I don't think it makes sense for us to pick this up.

**Comment by gvardany** [comment]
I will keep the issue in mind and revisit it in the proper context. I think we can close this PR without merging.
